lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -311,95 +298,49 @@ define((require, exports, module) ->
-        #      current token is null, and after stepping forward `tokenIterator` is on the
-        #      start of a document. If start sequence happens to be there, then equation start is
-        #      then found without any problem, whereas in this case null should be returned
-        for boundarySequence in boundarySequences
-          for token in boundarySequence.slice(0).reverse()
-            curToken = tokenIterator.getCurrentToken()
-            if curToken? and equalTokens(token, curToken)
-              moveFromBoundary()
-
-      tokenSequenceFinder = new TokenSequenceFinder(boundarySequences)
-      while true
-        moveToBoundary()
-        curToken = tokenIterator.getCurrentToken()
-        if not curToken?
-          return null
-
-        tokenSequenceFinder.progressIndices(curToken)
-
-        maybeFinishedSequenceId = tokenSequenceFinder.getMaybeFinishedSequenceId()
-        if maybeFinishedSequenceId?
-          { row: curTokenRow, column: curTokenColumn } = tokenIterator.getCurrentTokenPosition()
-          curTokenLength = curToken.value.length
-          finishedSequence = boundarySequences[maybeFinishedSequenceId]
-
-          equationOuterBoundary = {
-            row: curTokenRow
-            column: curTokenColumn + (if start then 0 else curTokenLength)
-          }
-
-          # after finding boundary, tokenIterator is exactly on the last token
-          # of ending sequence, so we step back until we're on the last token
-          # to determine inner boundary
-          for i in [0...finishedSequence.length-1]
-            moveFromBoundary()
-
-          { row: curTokenRow, column: curTokenColumn } = tokenIterator.getCurrentTokenPosition()
-          curTokenLength = tokenIterator.getCurrentToken().value.length
-
-          equationInnerBoundary = {
-            row: curTokenRow
-            column: curTokenColumn + (if start then curTokenLength else 0)
-          }
-
-          # to move inside the sequence, we move token iterator one more time
-          moveFromBoundary()
-
-          return {
-            id: maybeFinishedSequenceId
-            outer: equationOuterBoundary
-            inner: equationInnerBoundary
-          }
+      currentToken = tokenIterator.getCurrentToken()
+      { row: prevRow } = tokenIterator.getCurrentTokenPosition()
+      # TODO: magic string? importing is hard though
+      boundaryCorrect = true
+      while LatexParsingContext.isType(currentToken, "equation")
+        currentToken = moveToBoundary()
+        if not currentToken?
+          boundaryCorrect = false
+          break
+        { row: currentRow } = tokenIterator.getCurrentTokenPosition()
+        # Empty string always means that equation state is popped from state stack.
+        # Unfortunately, empty string is not tokenized at all, and TokenIterator
+        # just skips it altogether, so we have to handle this manually here.
+        if Math.abs(currentRow - prevRow) > 1
```

Comment:
Let's report empty string to the user somehow? Currently he just gets no preview and may think it is our fault; in fact it is _his_ fault and we may want to inform him: _hey guy, blank lines are not allowed in maths!_